src/clib/initialize_chemistry_data.c

```diff
@@ -435,6 +474,9 @@ int _free_chemistry_data(chemistry_data *my_chemistry,
     GRACKLE_FREE(my_rates->n_cr_n);
     GRACKLE_FREE(my_rates->n_cr_d1);
     GRACKLE_FREE(my_rates->n_cr_d2);
+    GRACKLE_FREE(my_rates->h2dustS);
+    GRACKLE_FREE(my_rates->h2dustC);
+    GRACKLE_FREE(my_rates->grain_growth_rate);
   }
```

Comment:
I just want to point out that we never free all of the new rate pointers. In other words, there is a memory leak. Given that most simulation codes usually just call ``initialize_chemistry_data`` once, this probably isn't an issue in most use-cases. But, this is something we need to address 

This is a massive PR and I don't want to burden you with too many extra things. So maybe it's okay to merge this PR without them? But if we do this, I think we need to hold of on releasing a new Grackle version that includes changes from this PR until we address this (or add clear documentation of this memory leak (with  ``primordial_chemistry==4``)

Reply:
It'd be safe we free all the new pointers before releasing a new grackle version.
I'll double check if we free all pointers.